### PR TITLE
Spatial parameters, diffusion units, initial confinement of species to subregion

### PR DESCRIPTION
**Pull request by emmetfrancis**
Unfortunately, due to some haphazard development of late, I have three new features introduced here. These are useful for some of the smart-comp-sci examples (can simplify some syntax, for instance).

1. Spatially varying parameters: We can now define species as spatially varying using the key symbols x, y, and z when initializing a parameter using `Parameter.from_expression()`. Temporal dependence was already implemented. For instance, to introduce a reaction rate than weakens linearly in the z direction (from $0 \leq z \leq L$) and fades out exponentially over time: `r1 = smart.model_assembly.Parameter.from_expression("r1", "(1-z/L)*exp(-t/1.0)", 1/unit.sec)`. This is utilized in a new iteration of example 3, `example3_withmanufactured`. Spatially-dependent parameters are defined and used in surface and volume reactions such that the solution takes a known analytical form for convergence testing.
2. Diffusion units fix: There was previously no check ensuring that the units of the diffusive terms in the variational form matched the units of other terms. Now, if the units do not match, the correct conversion is carried out.
3. Confine species to subregion: In certain examples (especially in smart-comp-sci, e.g. NMDAR in the dendritic spine example), species are initially (or always) confined to a subregion of a certain domain. This could be implemented by defining a new compartment in this region, but this would require repeating definitions of all species and reactions that appear in both subregions. In the dendritic_spine example, this is implemented by manually altering the initial condition of NMDAR, which involves some confusing syntax. Now, users can simply confine species to a specified region by using a mesh function with this region labeled (`mf_label`) with value `mf_val` and calling (for a species `sp`): `sp.restrict_to_subdomain(mf_label, mf_val)`
4. Other minor changes - Alter `implicit_curve` in `smart.mesh_tools` to account for maximum dimension in both the r and the z direction (used in smart-comp-sci repo examples), do not reverse smart objects (reactions, parameters, compartments, species) after adding them to the appropriate container, but maintain ordering. Also change `df[i][j]` indexing to `df[i].iloc[j]`.

### CODE REVIEW

File: examples/example3/example3_withmanufactured.ipynb [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+    "import numpy as np\n",
+    "import pathlib\n",
+    "import logging\n",
+    "\n",
+    "from smart import config, mesh, model, mesh_tools\n",
+    "from smart.units import unit\n",
+    "import sympy as sym\n",
+    "import gmsh  # must be imported before pyvista if dolfin is imported first\n",
+    "\n",
+    "from smart import config, common, mesh, model, mesh_tools, visualization\n",
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]:
```suggestion
    "from smart import config, mesh, model, mesh_tools, visualization\n",
```
Remove unused import.
Currently `visualization` is not used, (as you have commented out these plots), is that on purpose?

File: examples/example3/example3_withmanufactured.ipynb [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+    "    Reaction,\n",
+    "    Species,\n",
+    "    SpeciesContainer,\n",
+    "    ParameterContainer,\n",
+    "    CompartmentContainer,\n",
+    "    ReactionContainer,\n",
+    ")\n",
+    "\n",
+    "from matplotlib import pyplot as plt\n",
+    "import matplotlib.image as mpimg\n",
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb:56]:
```suggestion
```
unused import

File: examples/example3/example3_withmanufactured.ipynb [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+    "from matplotlib import rcParams"
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]:
```suggestion
```
unused import

File: examples/example3/example3_withmanufactured.ipynb [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+    "```\n",
+    "where\n",
+    "- name: string naming the species\n",
+    "- initial_condition: initial concentration for this species (can be an expression given by a string to be parsed by sympy - the only unknowns in the expression should be x, y, and z)\n",
+    "- concentration_units: concentration units for this species (μM here)\n",
+    "- D: diffusion coefficient\n",
+    "- diffusion_units: units for diffusion coefficient (μm<sup>2</sup>/sec here)\n",
+    "- compartment_name: each species should be assigned to a single compartment (\"Cyto\", here)\n",
+    "- group (opt): for larger models, specifies a group of species this belongs to;\n",
+    "            for organizational purposes when there are multiple reaction modules"
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]:
```suggestion
    "In this case, we have a single species, \"A\", which exists in the cytosol."
```
We already have this written down in other demos, so I don't see the need for it here.

File: examples/example3/example3_withmanufactured.ipynb [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+    "            reaction string\n",
+    "- explicit_restriction_to_domain: string specifying where the reaction occurs; required if the reaction is not\n",
+    "                                  constrained by the reaction string (e.g., if production occurs only at the boundary,\n",
+    "                                  as it does here, but the species being produced exists through the entire volume)\n",
+    "- group (opt): string placing this reaction in a reaction group; for organizational purposes when there are multiple reaction modules\n",
+    "- flux_scaling (opt): in certain cases, a given reactant or product may experience a scaled flux (for instance, if we assume that\n",
+    "                some of the molecules are immediately sequestered after the reaction); in this case, to signify that this flux \n",
+    "                should be rescaled, we specify ''flux_scaling = {scaled_species: scale_factor}'', where scaled_species is a\n",
+    "                string specifying the species to be scaled and scale_factor is a number specifying the rescaling factor"
+   ]
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]: Same comment here. I would only highlight new features in this demo here, and refer readers to other demos for more details (makes it easier to maintain and read).

File: examples/example3/example3_withmanufactured.ipynb [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+    "\n",
+    "which will serve as our parent mesh, giving the overall cell geometry.\n",
+    "\n",
+    "Different domains can be specified within this parent mesh by assigning marker values to cells (3D) or facets (2D) within the mesh. A subdomain within the parent mesh, defined by a region which shares the same marker value, is referred to as a child mesh.\n",
+    "\n",
+    "Here, we have two child meshes corresponding to the 2 compartments specified in the compartment container. As defined above, \"PM\" is a 2D compartment defined by facets with marker value 10 and \"Cyto\" is a 3D compartment defined by cells with marker value 1. These subdomains are defined by:\n",
+    "- $\\Omega_{Cyto}: r \\in [0, 1) \\subset \\mathbb{R}^3$\n",
+    "- $\\Gamma_{PM}: r=1 \\subset \\mathbb{R}^3$\n",
+    "\n",
+    "We generate the parent mesh with appropriate markers using gmsh in the function `mesh_tools.create_spheres`"
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]: Text out of date with actual code, uses box not sphere. Also remove any commented (unused code, such as create spheres in the code block below).

File: examples/example3/example3_withmanufactured.ipynb [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+    "    name=\"parent_mesh\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "309e5e8a",
+   "metadata": {},
+   "source": [
+    "Define manufactured solution and associated reactions. To not make the choice of manufactured solution completely arbitrary, we have chosen it as a multiplication of three solutions to the PDE in 1D, then multiplied by some time-dependence governed by the time-scale of protein dephosphorylation in the cytosol."
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]: We should probably have these equations in Latex format here as well, to make it easy to show the reader how to go from mathematics to reactions in smart.

File: examples/example3/example3_withmanufactured.ipynb [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+    "Ax = (k_kin*xSize/(2*D))*np.exp(xFactor) / ((xFactor/2)*(np.exp(xFactor)-1) + (k_kin*xSize/(2*D))*(1+np.exp(xFactor)))\n",
+    "Bx = (k_kin*xSize/(2*D)) / ((xFactor/2)*(np.exp(xFactor)-1) + (k_kin*xSize/(2*D))*(1+np.exp(xFactor)))\n",
+    "yFactor = ySize / np.sqrt(3*D/k_p)\n",
+    "Ay = (k_kin*ySize/(2*D))*np.exp(yFactor) / ((yFactor/2)*(np.exp(yFactor)-1) + (k_kin*ySize/(2*D))*(1+np.exp(yFactor)))\n",
+    "By = (k_kin*ySize/(2*D)) / ((yFactor/2)*(np.exp(yFactor)-1) + (k_kin*ySize/(2*D))*(1+np.exp(yFactor)))\n",
+    "zFactor = zSize / np.sqrt(3*D/k_p)\n",
+    "Az = (k_kin*zSize/(2*D))*np.exp(zFactor) / ((zFactor/2)*(np.exp(zFactor)-1) + (k_kin*zSize/(2*D))*(1+np.exp(zFactor)))\n",
+    "Bz = (k_kin*zSize/(2*D)) / ((zFactor/2)*(np.exp(zFactor)-1) + (k_kin*zSize/(2*D))*(1+np.exp(zFactor)))\n",
+    "expFactor = np.sqrt(k_p/(3*D))\n",
+    "Asol = cT * ((Ax*sym.exp(-expFactor*x) + Bx*sym.exp(expFactor*x)) * \n",
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]: There is a mixture of `numpy.exp` and `sympy.exp` here, which might confuse the reader, is there any particular reason we cannot use just `sympy` here?

File: examples/example3/example3_withmanufactured.ipynb [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "b432d407",
+   "metadata": {},
+   "source": [
+    "## Initialize model and solver\n",
+    "\n",
+    "Now we modify the solver configuration for this problem. In the solver config, we set the final t as 1 s, the initial dt at .01 s (without any additional specifications, this will be the time step for the whole simulation), and the time precision (number of digits after the decimal point to round to) as 6."
```
Comment (jorgensd) [inline examples/example3/example3_withmanufactured.ipynb]:
```suggestion
    "Now we modify the solver configuration for this problem. In the solver config, we set the simulation end time (`\"final_t\"`) as 1 s, the initial time step (`\"initial_dt\"`) at .01 s (without any additional specifications, this will be the time step for the whole simulation), and the time precision (number of digits after the decimal point to round to) as 6."
```

File: smart/mesh_tools.py
Comment (jorgensd) [inline smart/mesh_tools.py:1]: This is not used in the code above, and is only called in: 
`create_2Dcell` and `create_axisymm` (never called in library).
Should we move this to a separate PR?

File: smart/model.py [hunk 10]
```diff
@@ -975,6 +980,34 @@ def _init_4_7_set_initial_conditions(self):
                     self.dolfin_set_function_values(
                         species, ukey, species.initial_condition_expression
                     )
+                if species.has_subdomain:
+                    # restrict to specified subdomain
+                    uvec = self.cc[species.compartment_name].u["u"].vector()
+                    values = uvec.get_local()
+                    values_new = np.zeros(len(values))
```
Comment (jorgensd) [inline smart/model.py]:
```suggestion
                    values_new = np.zeros_like(values)
```

File: smart/model.py [hunk 11]
```diff
@@ -975,6 +980,34 @@ def _init_4_7_set_initial_conditions(self):
...
+                                local_idx = local_idx[0][0]
+                                cur_sub_idx = d.vertex_to_dof_map(funcSpace)[local_idx]
+                                if cur_marker == species.subdomain_val:
+                                    values_new[species.dof_map[cur_sub_idx]] = values[
+                                        species.dof_map[cur_sub_idx]
+                                    ]
+                    values[species.dof_map] = values_new[species.dof_map]
+                    vec = self.cc[species.compartment_name].u[ukey].vector()
+                    vec.set_local(values)
+                    vec.apply("insert")
```
Comment (jorgensd) [inline smart/model.py]: Is there a reason we do this and not interpolation? Interpolation would be way faster (and should work with the latest fixes in meshview).

File: smart/model.py [hunk 12]
```diff
@@ -1040,17 +1083,17 @@ def _init_5_2_create_variational_forms(self):
                     extra=dict(format_type="log"),
                 )
             else:
+                if Dform_units != mass_form_units:  # unit conversion for consistency
+                    diffusion_conversion = species.diffusion_units.to(
+                        species.compartment.compartment_units**2 / unit.s
+                    )
+                    D = D * diffusion_conversion.magnitude
```
Comment (jorgensd) [inline smart/model.py]:
```suggestion
                    D *= diffusion_conversion.magnitude
```

File: smart/model_assembly.py
Comment (jorgensd) [inline smart/model_assembly.py:1]: Some unrelated changes in here as well. Would be good to either:
- Put them in  a separate PR
or
- Explain the changes in the PR text.

File: examples/example3/example3_withmanufactured.ipynb (hunk 4 above)
Comment (emmetfrancis) [reply examples/example3/example3_withmanufactured.ipynb]: Forgot to remove and update text in this reworked example, apologies!

File: examples/example3/example3_withmanufactured.ipynb (hunk 8 above)
Comment (emmetfrancis) [reply examples/example3/example3_withmanufactured.ipynb]: Good catch, should just be fine as just `sympy.exp` I think.

File: smart/mesh_tools.py
Comment (emmetfrancis) [reply smart/mesh_tools.py:1]: > This is not used in the code above, and is only called in: `create_2Dcell` and `create_axisymm` (never called in library). Should we move this to a separate PR?

This function was already present, these were just a couple minor changes that may have made their way in from another project. I can revert here and open a new PR with an example that uses this function, if that sounds best.

Comment (jorgensd) [reply smart/mesh_tools.py:1]: Im fine with keeping them if you mention them in the PR description:)

File: smart/model.py (hunk 11 above)
Comment (emmetfrancis) [reply smart/model.py]: I have altered this section of the code now to only iterate over facets in the subdomain of interest. I may be missing something obvious, but I could not determine a way to use `dolfin.interpolate` to interpolate selectively onto a subregion of a MeshView-defined function. The desired purpose of this function is to keep the function `self.cc[species.compartment_name].u["u"]` in the same function space, but set its values to zero everywhere outside of some subregion specified by a MeshFunction.

File: smart/model_assembly.py
Comment (emmetfrancis) [reply smart/model_assembly.py:1]: Some changes were reverted to be included in the Latex tables PR, the others are now explained in point 4 of the PR description above.

File: smart/model.py (hunk 11 above)
Comment (jorgensd) [reply smart/model.py]: To summarize what you have here:
1. A parent mesh (from now called `p_mesh`) that is 2D
2. A mesh-view (from now called `s_mesh`) that is 1D and a subset of the facets of the parent mesh

What you want to do is to take a function `u_s` that is defined on `s_mesh`, and restrict the function to a subset of facets in `p_mesh` (and set it to zero everywhere else). However, the function `u_s_new` is still defined on `s_mesh`?
Could you confirm that I understand you correctly?

Comment (emmetfrancis) [reply smart/model.py]: That is correct. The meshview `s_mesh` could be the same dimension as the parent mesh and this could all be in 3D/2D as well, though.

Comment (jorgensd) [reply smart/model.py]: @emmetfrancis  Here is an example of doing the same coding, but without looping through vertices (would generalize to higher order spaces).

```python
from dolfin import *
import numpy as np



def create_restriction(u, mesh_function, value):
    """
    Restrict a function on a submesh to a subset of parent entities (same dimension as the submesh)

    :param u: Function on submesh
    :param mesh_function: MeshFunction marking the subset of parent entities (same dimension as the cells of the submesh)
    :param value: Value in MeshFunction marking the subset of parent entities
    :return: New restricted function
    """
    submesh = u.function_space().mesh()

    # Compute local cells in submesh marked by parent meshtag
    sub_to_parent_map = submesh.topology().mapping()[mesh_function.mesh().id()].cell_map()
    marked_sub_entities = mesh_function.array()[sub_to_parent_map] == value
    local_indices = np.flatnonzero(marked_sub_entities)

    # Find all degrees of freedom to transfer data from
    V = u.function_space()
    u_new = Function(u.function_space())
    vector = u_new.vector()
    dof_list = [V.dofmap().cell_dofs(cell) for cell in local_indices]
    if len(dof_list) == 0:
        transfer_dofs = np.array([])
    else:
        transfer_dofs = np.unique(np.hstack(dof_list))
    im = V.dofmap().index_map()
    num_local = im.local_range()[1] - im.local_range()[0]

    # Filter out dofs that are not local
    transfer_dofs = np.array([dof for dof in transfer_dofs if dof < num_local])
    vector[transfer_dofs] = u.vector()[transfer_dofs]
    vector.apply('insert')
    u_new.rename("u_new", "u_new")

    return u_new


class Left(SubDomain):
    def inside(self, x, on_boundary):
        return x[0] < 0.5

class LeftLeft(SubDomain):
    def inside(self, x, on_boundary):
        return x[0] <= 0.25 + 1e-14


def create_example(tdim, codim):
    if tdim == 2:
        mesh = UnitSquareMesh(20, 10)
    elif tdim == 3:
        mesh = UnitCubeMesh(10, 5, 4)

    value = 7
    restriction_value = 5
    mf = MeshFunction("size_t", mesh, tdim-codim, 0)
    Left().mark(mf, value)

    mf2 = MeshFunction("size_t", mesh, tdim-codim, 0)
    LeftLeft().mark(mf2, restriction_value)

    submesh = create_meshview(mf, value)




    V = FunctionSpace(submesh, 'CG', 1)
    u = Function(V)
    u.interpolate(Expression("x[0]+x[1]", degree=1))
    
    u_new = create_restriction(u, mf2, restriction_value)
    with XDMFFile(f"u_new_{tdim}_{codim}.xdmf") as file:
        file.write(u_new)



if __name__ == "__main__":
    create_example(2, 0)
    create_example(2, 1)
    create_example(3, 0)
    create_example(3, 1)
```

Comment (emmetfrancis) [reply smart/model.py]: Thank you, this looks great. I will incorporate and test it out asap.

**Review by finsberg — APPROVED** [review]
I think we should merge this one and then follow up with additional changes in a new PR if needed.
